src/background.ts

```diff
@@ -337,6 +342,14 @@ ipcMainHandle("RESTART_ENGINE", async () => {
   });
 });
 
+ipcMainHandle("CHANGE_PIN_WINDOW", () => {
```

Comment:
関数を実行するたびにtrue/falseを入れ替えるのではなく、`useGpu`のようにするのはどうでしょう。
https://github.com/Hiroshiba/voicevox/blob/43d267b7fe21da9840a47e8c84a0c1fcc44203c0/src/background.ts#L280
こうすれば外部要因でピン状態が変更された場合でも正常に稼働すると思います。
（vuexのstore以外になるべく状態を持たせたくない、という意図もあります。）

---

Instead of swapping true/false each time the function is executed, how about doing something like `useGpu`?
https://github.com/Hiroshiba/voicevox/blob/43d267b7fe21da9840a47e8c84a0c1fcc44203c0/src/background.ts#L280
If you do this, it will work correctly even if the pin state is changed by external systems.
(It is also our intention not to have any state other than vuex's store.)

Reply:
I think it's not a `vuex` state problem, I implemented it following the maximize button, which goes through a route like
```
button --window.electron.changePinWindow()--> ipcRenderer --invoke() -> ipcHandler -> window state

window state --win.on("always-on-top-changed")-- --win.webContents.send() --> ipc message receiver --> vuex state --> button
                               /\
       however this part can not be triggered by external program
```
The problem happens on where `win.on("always-on-top-changed")` is not called when it's set always on top by an external program, which is also the only way we can get to know it. I think it's a total electron bug but not a vuex's or something.

---

そうでない気がします。ピンボタンの仕組みがほとんど最大化ボタンと同じ、こうしています：
```
button --window.electron.changePinWindow()--> ipcRenderer --invoke() -> ipcHandler -> window state

window state --win.on("always-on-top-changed")-- --win.webContents.send() --> ipc message receiver --> vuex state --> button
                               /\
     　　　　　　　　　　　　  この部分はバグ
```

結果としてこれはただの`win.on("always-on-top-changed")`が呼び出さないelectronの問題ですから、vuexで改善できる気がしないんです。

Reply:
おっと、認識に違いがあるかもしれません。

仰るとおり、vuexの状態が（`always-on-top-changed`バグにより）実態（`win.isAlwaysOnTop()`）と乖離するはずです。
このときに`CHANGE_PIN_WINDOW`を実行すると、ピンボタンのマークと実態が逆になったままになります。

```
vuex.isPinned: TRUE　isAlwaysOnTop(): TRUE

↓　bug!

vuex.isPinned: TRUE　isAlwaysOnTop(): FALSE　（これは仕方ない。）

↓　`CHANGE_PIN_WINDOW`

vuex.isPinned: FALSE　isAlwaysOnTop(): TRUE　（これを防ぎたい。）
```

なので、true/falseを反転させる`CHANGE_PIN_WINDOW`ではなく状態をコピーする`SET_IS_PINNED_WINDOW`などにするのはどうでしょう。

私のイメージはこんな感じの関数です。
```ts
"SET_IS_PINNED_WINDOW": (isPinned)=>{
  if (win.isAlwaysOnTop() != isPinned) { win.setAlwaysOnTop(isPinned); }
}
```

---

Oops, there seems to be a difference in perception.

As you said, the state of vuex (due to the `always-on-top-changed` bug) should be different from the actual state (`win.isAlwaysOnTop()`).
If you execute `CHANGE_PIN_WINDOW` in this case, the pin button mark and the actual state will remain reversed.

```
vuex.isPinned: TRUE isAlwaysOnTop(): TRUE

↓ bug!

vuex.isPinned: TRUE　isAlwaysOnTop(): FALSE　(This can't be avoided.)

↓ `CHANGE_PIN_WINDOW`.

vuex.isPinned: FALSE　isAlwaysOnTop(): TRUE　(I want to prevent this.)
```

So, instead of flipping true/false, how about copying the state?

I'm imagining a function like this.
```ts
"SET_IS_PINNED_WINDOW": (isPinned)=>{
  if (win.isAlwaysOnTop() != isPinned) { win.setAlwaysOnTop(isPinned); }
}
```

Reply:
No, it's even worse. Passing the value to background causes another bug when the button is at unpinned and an external program set it to be always on top.

Under such circumstance, the unpinned button will always send `true` to background and since window status is already set to be true, it will never call a `win.on("always-on-top-changed")` to update `vuex` status and the whole system will be stuck there forever.

On the other hand, the current function reverses the window status every time it's called, which makes sure the `win.on("always-on-top-changed")` is always triggered at a click on pin button. It then update the `vuex` status to sync with the window status, so that the bug only causes necessity of another click to reset the status.

This doesn't actually result into the situation you described, for `vuex` status is synced with window status through `win.on("always-on-top-changed")`.

The point is that although the route
```
button --window.electron.changePinWindow()--> ipcRenderer --invoke() -> ipcHandler -> window state
```

reverses the window status referring to previous status, the way back from background

```
window state --win.on("always-on-top-changed")-- --win.webContents.send() --> ipc message receiver --> vuex state --> button
```

carries a valid data of window status.

So the status is actually like this:

```
vuex.isPinned: TRUE　isAlwaysOnTop(): TRUE

↓　bug!

vuex.isPinned: TRUE　isAlwaysOnTop(): FALSE　（これは仕方ない。）

↓　`CHANGE_PIN_WINDOW`

vuex.isPinned: FALSE　isAlwaysOnTop(): TRUE　（Ths change will trigger win.on("always-on-top-changed")）

↓　`DETECT_PINNED`(carries the isAlwaysOnTop: TRUE)

vuex.isPinned: TRUE　isAlwaysOnTop(): TRUE
```

I don't think we need to add anything more here.

---

いいえ、むしろこうすると悪化しまうかもしれません。ボタンがアンピンの状態で外部プログラムがウインドウを最前面に表示にすると、バグが出てきます。

その状態では、ボタンは常に`true`を送っていますけど、バックグラウンドではもう`true`だから、 `win.on("always-on-top-changed")`は呼び出されません。この関数は同時に`vuex`の`isPinned`を更新するので、このままボタンも更新せず詰みます。

ところが、今ではボタンの状態を問わず毎回ウインドウの状態を反転するから、`win.on("always-on-top-changed")`も無事で呼び出し、`vuex`を更新していく仕組みです。結果として、もう一回でボタンをクリックしたら最初の状態に戻れます。

この仕組みでは、ご心配の状況にならない自信があります。なせなら、このルート

```
button --window.electron.changePinWindow()--> ipcRenderer --invoke() -> ipcHandler -> window state
```

はボタンの状態を伝わらないんですが、バックグラウンドから戻るルート

```
window state --win.on("always-on-top-changed")-- --win.webContents.send() --> ipc message receiver --> vuex state --> button
```

はちゃんとウインドウの状態で`vuex`を更新しますから。

そして実際の状態はこうなります：

```
vuex.isPinned: TRUE　isAlwaysOnTop(): TRUE

↓　bug!

vuex.isPinned: TRUE　isAlwaysOnTop(): FALSE　（これは仕方ない。）

↓　`CHANGE_PIN_WINDOW`

vuex.isPinned: FALSE　isAlwaysOnTop(): TRUE　（この時`win.on("always-on-top-changed")`も呼び出されます）

↓　`DETECT_PINNED`(isAlwaysOnTop: TRUEを示している)

vuex.isPinned: TRUE　isAlwaysOnTop(): TRUE
```

これ以上加える必要がない気がします。

Reply:
なるほど、あなたの意見は筋が通っているように思いました。

ちょっと整理してみます。
私たちの意見が違っているのは、`always-on-top-changed`のバグによりwinとvuexで同じ状態を持つことができないためです。
同じ状態を持てないので、どちらか一方を信頼し、もう一方にコピーする必要があります。

あなたはwinの状態を信頼するべきだとしています。おそらく、外部システムによってwinの状態は正確に変わる一方、それをvuexにコピーできないためでしょう。
一方私はvuexの状態を信頼するべきだとしています。これは、他のすべての状態はvuexに集約されているので、ピンの状態もvuexのものを信頼するほうが実装の見通しが良いからです。

今回はwinの状態を信頼するというあなたの実装にして良いと思いました。
そして将来バグが直ったときに、状態をvuexに集約することをもう一度考えるというのはどうでしょうか？

---

Okay, your opinion seemed to make sense.

Let me try to sort this out a bit.
The reason we have different opinions is that we can't have the same state in win and vuex due to the `always-on-top-changed` bug.
Since we can' t have the same state, we need to trust one and copy it to the other.

You are saying that we should trust the win state. Probably because the state of win changes exactly depending on the external system, but it can' t be copied to vuex.
On the other hand, I am saying that we should trust the state of vuex. This is because all the other states are aggregated in vuex, so it is a better implementation prospect to trust the pin state in vuex as well.

In this case, I thought your implementation of trusting the win state is a good idea.
And in the future, when the bug is fixed, how about thinking again about aggregating the state to vuex?

Reply:
Sure, let's put off this discussion till the electron bug get fixed, we've already spent too much time on this tiny feature.

---

ええ、そうしましょう。この機能はとても小さいのにかなりの時間がかかったので、この討論はelectron側がバグを直ったときまで一旦止めるほうがいいと思います。